docs/source/jobs.rst

```diff
@@ -239,41 +239,143 @@ Use cases for the **job document** include, but are not limited to:
+
+Reading and Writing data
+------------------------
 
 An example of storing data:
 
 .. code-block:: python
 
     >>> import numpy as np
     >>> job = project.open_job(statepoint)
-    >>> job.data['x'] = np.arange(0, 1, 0.01)
-    >>> job.data['hello'] = 'world'
+    >>> job.data['x'] = np.ones([10, 3, 4])
+
 
 Just like the job *state point* and *document*, individual keys may be accessed either as attributes or through a functional interface, *e.g.*.
-The following examples are all equivalent:
+
+To access data as an attribute:
 
 .. code-block:: python
 
-    >>> print(job.data.get('hello'))
-    world
-    >>> print(job.data['hello'])
-    world
-    >>> print(job.data.hello)
-    world
+    >>> with job.data:
+    ...     x = job.data.x[:]
+
+To access data as a key:
+
+.. code-block:: python
+
+    >>> with job.data:
+    ...     x = job.data['x'][:]
+
+Through a functional interface:
+
+.. code-block:: python
+    
+    >>> with job.data:
+    ...     x = job.data.get('x')[:]
 
 .. tip::
 
-     Use the :py:meth:`Job.data.get` method to return ``None`` or another specified default value for missing values. This works exactly like with python's `built-in dictionaries <https://docs.python.org/3/library/stdtypes.html#dict.get>`_.
+     Use the :py:meth:`Job.data.get` method to return ``None`` or another specified default value for missing values. This works exactly like with python's built-in dictionaries (see :py:meth:`dict.get`). 
+    
+Accessing arrays
+----------------
+
+All values stored in :attr:`job.data` are returned as copies, except for arrays, which are accessed *by reference* and not automatically copied into memory.
+That is important to enable the storage of massive arrays that do not necessarily fit into memory.
+
+For fast and efficient data access, NumPy slicing syntax may be used to access data. 
+Here are a few examples for accessing a three-dimensional array with outputs omitted:
+   
+.. code-block:: python
+    
+    >>> with job.data:
+    ...     job.data.x[0, 0, 0]
+    ...     job.data.x[1:3, 0, :]
+    ...     job.data.x[:, 1, 3]
+    
+To load entire arrays to memory, NumPy slicing syntax may be used:
+ 
+.. code-block:: python
+
+    >>> with job.data:
+    ...     x = job.data.x[:]
+
+NumPy slicing (ie. the ``[:]`` operator) may be used to load array-like and text data.
+It cannot be used to load scalar values.
+Instead, the explicit memory copy operator ``[()]`` may be used instead of NumPy slicing to load entire arrays or scalars to memory:
+
+.. code-block:: python
+
+    >> with job.data:
+    ..      x = job.data.x[()]
+
+A caveat of the explicit memory copy operator ``[()]`` is that it cannot be used to load strings.
+Generally, the :py:attr:`job.data` container is intended for large numerical or text data.
+Information which needs to be searchable and scalars should be stored in the :ref:`job document <project-job-document>`.
```

Comment:
I don't think we should suggest that scalars should *generally* be stored in the job-doc, however, anything that should be searchable *must* be stored in the job-doc, so I suggest this minor adjustment:
```suggestion
Information which needs to be searchable, typically scalars or smaller text-like data, should be stored in the :ref:`job document <project-job-document>`.
```